File: LeetCode/Problems/Python/621. Task Scheduler.py
#Time: O(n)
#Space: O(1), 26 characters in total
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        if not tasks:
            return 0
        
        count = list(Counter(tasks).values())
        time = 0
        count.sort()
        max_task =count.pop()
        max_idle = (max_task-1) * n
        while count and max_idle > 0:
            max_idle-= min(max_task-1, count.pop())
            
        max_idle = max(0, max_idle)
        return max_idle + len(tasks)
                

File: LeetCode/Problems/C++/621. Task Scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        if(!tasks.size()) return 0;
        
        int count[26]{};
        int max_idle;
        int ptr = 24;
        for(int i=0;i<tasks.size();i++) {
            count[tasks[i]-'A']+=1;
        }
        sort(count, count+26);
        
        max_idle = (count[25]-1) * n;
        
        while(ptr>=0 && max_idle >= 0)
            max_idle-=min(count[25]-1,count[ptr--]);
        
        max_idle = max(0, max_idle);
        
        return max_idle + tasks.size();    
    }
};